### PR TITLE
autorotate in-memory jpegs

### LINKED ISSUE #8172: Error in imdecode, acts differently than imread and doesn't things like orientation and more correctly
##### System information (version)
- OpenCV => 3.1.0
- Operating System / Platform => Ubuntu 16.04.1 LTS
- Compiler => gcc version 5.4.0 20160609

##### Detailed description

There appears to be a bug in python cv2.imdecode behaves differently than cv2.imread, imread will correctly detect orientation and load the image properly, meanwhile imdecode will neither detect orientation or load the image properly. I cannot upload the input image itself as it contains sensitive data, but it's a completely standard image from an iPhone 6 with the following meta data (from file), displaying orientation etc:

`JPEG image data, Exif standard: [TIFF image data, big-endian, direntries=10, manufacturer=Apple, model=iPhone 6, orientation=upper-right, xresolution=150, yresolution=158, resolutionunit=2, software=10.0.1, datetime=2016:12:01 15:19:20], baseline, precision 8, 3264x2448, frames 3`

##### Steps to reproduce

Load the file like:

```
fh = open('somefile.jpg', 'r') # or some other way of providing in memory streams, i.e. posted from a webserver
arr = np.asarray(bytearray(fh.read()), dtype=np.uint8)
tmp = cv2.imdecode(arr, -1) # 'load it as it is' or 1 for 3channel
cv2.imwrite('/tmp/loadedfile.jpg', tmp)
```

This will write the image back incorrectly to the disk, meanwhile if i either write the memory stream to disk and then use imread, it handles orientation and so forth correctly.

### PR DESCRIPTION
**Pull request by brian-armstrong-discord**
resolves #8172

This pull request changes ExifReader to operate on istreams rather than filenames. It then builds a streambuf on top of Mat so that imdecode may use ApplyExifOrientation like imread does.

### CODE REVIEW

File: modules/imgcodecs/src/loadsave.cpp [hunk 1]
```diff
@@ -284,6 +295,51 @@ static void ApplyExifOrientation(const String& filename, Mat& img)
...
+}
+
+static void ApplyExifOrientation(const Mat& buf, Mat& img)
+{
+    int orientation = IMAGE_ORIENTATION_TL;
+
+    Size bufSize = buf.size();
+    if( bufSize.width == 1 && buf.type() == CV_8UC1 )
+    {
+       ByteStreamBuffer bsb( reinterpret_cast<char*>(buf.data), bufSize.height );
```
Comment (brian-armstrong-discord) [inline modules/imgcodecs/src/loadsave.cpp]: ahhh ok that makes sense 👍 i still want to check for `CV_8UC1` though, right?

**Comment by brian-armstrong-discord** [comment]
The test that failed here doesn't seem related to this patch

`OpenCV: FFMPEG: tag 0x34363248/'H264' is not supported with codec id 28 and format 'mp4 / MP4 (MPEG-4 Part 14)'
OpenCV: FFMPEG: fallback to use tag 0x00000021/'!???'`

Could you help me find what I might have broke?

File: modules/imgcodecs/src/loadsave.cpp [hunk 2]
```diff
@@ -284,6 +306,50 @@ static void ApplyExifOrientation(const String& filename, Mat& img)
...
+}
+
+static void ApplyExifOrientation(const Mat& buf, Mat& img)
+{
+    int orientation = IMAGE_ORIENTATION_TL;
+
+    if( buf.isContinuous() )
+    {
+       ByteStreamBuffer bsb( reinterpret_cast<char*>(buf.data), buf.total() * buf.elemSize() );
+       std::istream stream( &bsb );
```
Comment (alalek) [inline modules/imgcodecs/src/loadsave.cpp]: Looks like the current `ByteStreamBuffer` implementation is buggy.

Please use this instead (with added `#include <sstream>` and removed `ByteStreamBuffer`):
```
std::istringstream stream(std::string(reinterpret_cast<char*>(buf.data), buf.total() * buf.elemSize()), std::ios_base::in);
```

(it does extra `memcpy` operation, but it is not critical for I/O tasks)

**Comment by brian-armstrong-discord** [comment]
@alalek Is there any way to retrigger the builds? I'm curious if this build error is going to happen consistently. The builds all passed previously, I think I barely made any changes after that

File: modules/imgcodecs/src/loadsave.cpp (hunk 2 above)
Comment (brian-armstrong-discord) [inline modules/imgcodecs/src/loadsave.cpp]: i'm not sure this is what caused build to break - it passed with a nearly identical implementation

**Comment by brian-armstrong-discord** [comment]
@alalek Do you feel pretty confident that's what caused it to break? The build did pass on a previous commit, though with the squash the changes have been hidden. IIRC the only things I changed were that I added an additional bounds check for `INT_MIN` and added your suggestion to use `buf.total() * bufSize.elemSize()`. That doesn't seem like enough to cause video files? to fail - in fact, after checking the test in question, I'm not even sure imread/imdecode are being invoked.

**Comment by alalek** [comment]
Stack trace for `ByteStreamBuffer` ctor call:
```
#0  0x00000000005a3ef3 in (anonymous namespace)::ByteStreamBuffer::ByteStreamBuffer(char*, std::size_t) (this=0x7ffffffecf50, base=0x1908bf0 "\377\330\377", <incomplete sequence \340>, length=38608) at /home/alalek/projects/opencv/opencv_dev/modules/imgcodecs/src/loadsave.cpp:66
#1  0x00000000005a5c97 in cv::ApplyExifOrientation(cv::Mat const&, cv::Mat&) (buf=..., img=...)
    at /home/alalek/projects/opencv/opencv_dev/modules/imgcodecs/src/loadsave.cpp:339
#2  0x00000000005a73c1 in cv::imdecode(cv::_InputArray const&, int) (_buf=..., flags=3)
    at /home/alalek/projects/opencv/opencv_dev/modules/imgcodecs/src/loadsave.cpp:733
#3  0x000000000058f720 in cv::MotionJpegCapture::retrieveFrame(int, cv::_OutputArray const&) (this=0x1908840, output_frame=...)
    at /home/alalek/projects/opencv/opencv_dev/modules/videoio/src/cap_mjpeg_decoder.cpp:824
#4  0x0000000000582bbb in cv::VideoCapture::retrieve(cv::_OutputArray const&, int) (this=0x7ffffffed4d0, image=..., channel=0)
    at /home/alalek/projects/opencv/opencv_dev/modules/videoio/src/cap.cpp:640
#5  0x0000000000582d96 in cv::VideoCapture::read(cv::_OutputArray const&) (this=0x7ffffffed4d0, image=...)
    at /home/alalek/projects/opencv/opencv_dev/modules/videoio/src/cap.cpp:663
#6  0x0000000000582dfb in cv::VideoCapture::operator>>(cv::Mat&) (this=0x7ffffffed4d0, image=...)
    at /home/alalek/projects/opencv/opencv_dev/modules/videoio/src/cap.cpp:691
#7  0x00000000004fc681 in CV_PositioningTest::run(int) (this=0x7fffffffd6f0)
    at /home/alalek/projects/opencv/opencv_dev/modules/videoio/test/test_video_pos.cpp:142
```

It is called from "MotionJpegCapture" in this test.
BTW, Probably it is better to use `IMREAD_IGNORE_ORIENTATION` flag in MJPEG decoder.

Crash backtrace (with [log messages](https://github.com/alalek/opencv/commit/f75528ec5410c043baf4dd29cec44f9ccba43a03)):
```
ByteStreamBuffer(0x18f2310, 28272)
ByteStreamBuffer::seekoff(14, 1, ...)
ByteStreamBuffer::seekoff(65, 1, ...)
ByteStreamBuffer::seekoff(65, 1, ...)
ByteStreamBuffer::seekoff(29, 1, ...)
ByteStreamBuffer::seekoff(179, 1, ...)
ByteStreamBuffer::seekoff(29, 1, ...)
ByteStreamBuffer::seekoff(179, 1, ...)
ByteStreamBuffer::seekoff(15, 1, ...)
ByteStreamBuffer::seekoff(10, 1, ...)
ByteStreamBuffer::seekoff(35366, 1, ...)

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff102d970 in __memcpy_sse2_unaligned () from /lib64/libc.so.6
(gdb) bt
#0  0x00007ffff102d970 in __memcpy_sse2_unaligned () at /lib64/libc.so.6
#1  0x00007ffff7b7833b in std::basic_streambuf<char, std::char_traits<char> >::xsgetn(char*, long) () at /lib64/libstdc++.so.6
#2  0x00007ffff7b51b9b in std::istream::read(char*, long) () at /lib64/libstdc++.so.6
#3  0x00000000005ca5c0 in cv::ExifReader::getExif() (this=0x7ffffffecf90)
    at /home/alalek/projects/opencv/opencv_dev/modules/imgcodecs/src/exif.cpp:134
#4  0x00000000005ca1ba in cv::ExifReader::parse() (this=0x7ffffffecf90) at /home/alalek/projects/opencv/opencv_dev/modules/imgcodecs/src/exif.cpp:83
#5  0x00000000005a5cd4 in cv::ApplyExifOrientation(cv::Mat const&, cv::Mat&) (buf=..., img=...)
    at /home/alalek/projects/opencv/opencv_dev/modules/imgcodecs/src/loadsave.cpp:342
#6  0x00000000005a73c1 in cv::imdecode(cv::_InputArray const&, int) (_buf=..., flags=3)
    at /home/alalek/projects/opencv/opencv_dev/modules/imgcodecs/src/loadsave.cpp:733
#7  0x000000000058f720 in cv::MotionJpegCapture::retrieveFrame(int, cv::_OutputArray const&) (this=0x1908840, output_frame=...)
    at /home/alalek/projects/opencv/opencv_dev/modules/videoio/src/cap_mjpeg_decoder.cpp:824
#8  0x0000000000582bbb in cv::VideoCapture::retrieve(cv::_OutputArray const&, int) (this=0x7ffffffed4d0, image=..., channel=0)
    at /home/alalek/projects/opencv/opencv_dev/modules/videoio/src/cap.cpp:640
#9  0x0000000000582d96 in cv::VideoCapture::read(cv::_OutputArray const&) (this=0x7ffffffed4d0, image=...)
    at /home/alalek/projects/opencv/opencv_dev/modules/videoio/src/cap.cpp:663
#10 0x0000000000582dfb in cv::VideoCapture::operator>>(cv::Mat&) (this=0x7ffffffed4d0, image=...)
    at /home/alalek/projects/opencv/opencv_dev/modules/videoio/src/cap.cpp:691
#11 0x00000000004fc681 in CV_PositioningTest::run(int) (this=0x7fffffffd6f0)
    at /home/alalek/projects/opencv/opencv_dev/modules/videoio/test/test_video_pos.cpp:14
```


The similar logs messages are on [buildbot](http://pullrequest.opencv.org/buildbot/builders/precommit_linux64_no_opt/builds/10464/steps/test_videoio/logs/stdio) (search for "Failure" keyword).

**Comment by brian-armstrong-discord** [comment]
Ah, that helps a lot.

I don't see stacktraces in the buildbot output. Maybe I don't have enough permissions? All I see is

```/build/precommit_linux64_no_opt/opencv/modules/ts/src/ts.cpp:540: Failure
Failed

	failure reason: Arithmetic exception
	test case #-1
	seed: ffffffffffffffff
-----------------------------------
	LOG:

File: /tmp/__opencv_temp.jNDKKz/__opencv_temp.YQfV7G.XVID.avi

File: /tmp/__opencv_temp.jNDKKz/__opencv_temp.UJvnn8.MPEG.avi

File: /tmp/__opencv_temp.jNDKKz/__opencv_temp.loKZRG.MJPG.avi
General failure:
	Arithmetic exception (-6)

-----------------------------------```

which didn't really tell me anything :)

**Comment by alalek** [comment]
`ByteStreamBuffer` has very limited implementation. Looks like it handles `dir == std::ios_base::cur` case only. Unfortunately there are no reliable checks for unsupported parameters (just several `return -1` statements for some cases).
So I still prefer to avoid custom implementation of mapped byte stream and to use `std::istringstream` instead.

**Comment by brian-armstrong-discord** [comment]
@alalek If you look at ExifReader.cpp, you can see that the only seeking it does (with `seekg`) is based on the current pointer, so in fact this is fine. If I changed ExifReader to inspect the return values from these `seekg` calls, would that work for you? It seems silly to me to allocate and copy a bunch of memory just for this purpose

**Comment by brian-armstrong-discord** [comment]
@alalek Also, for what it's worth, `seekoff`'s base behavior is to return -1, so doing so here merely conforms to that. http://www.cplusplus.com/reference/streambuf/streambuf/seekoff/

**Comment by brian-armstrong-discord** [comment]
@alalek I've pushed a new version which will check the result of `seekg`. I think this is in fact safer than the previous version which never checked the result of `seek`

File: modules/imgcodecs/src/loadsave.cpp [hunk 3]
```diff
@@ -50,10 +50,55 @@
...
+      {
+         if( gptr() + off >= egptr() )
+         {
+            return -1;
+         }
+         if( gptr() + off < eback() )
+         {
+            return -1;
+         }
+         gbump( static_cast<int>(off) );
```
Comment (alalek) [inline modules/imgcodecs/src/loadsave.cpp]: "int" type limitation can be dropped (with `INT_MAX`/`INT_MIN` checks above):
```
setg(eback(), gptr() + off, egptr());
```

Comment (brian-armstrong-discord) [inline modules/imgcodecs/src/loadsave.cpp]: it can't be dropped because the windows build complains about silently converting a long to an int

Comment (alalek) [inline modules/imgcodecs/src/loadsave.cpp]: I mean this:
```.diff
- gbump( static_cast<int>(off) );
+ setg(eback(), gptr() + off, egptr());
```
There is no conversion to "int".

Comment (brian-armstrong-discord) [inline modules/imgcodecs/src/loadsave.cpp]: 👍

**Comment by brian-armstrong-discord** [comment]
@alalek Thanks for all your help so far 👍 I know it's been kind of a struggle, heh

Let me know if there's anything else I can do to help this PR along

**Comment by alalek** [comment]
:+1:

**Comment by ndtreviv** [comment]
Forgive my idiocy, but which version did this ship in? I've got 3.3.1 and it's still not autorotating images based on exif rotation settings via imdecode...

**Comment by caogang** [comment]
This bug is fixed after which version?

**Pull request merged**